tests/unit/process_product_edit_rules.t

```diff
@@ -0,0 +1,215 @@
+			[{name => "Disallow ingredients", actions => [["ignore_if_0_nutriment_fruits-vegetables-nuts"]]},],
+		form => {"nutriment_fruits-vegetables-nuts" => 0},
+		result => 1,
+		delete_param => ["nutriment_fruits-vegetables-nuts"],
+	},
+	{
+		id => "ignore_if_equal_nutriments_sugar",
+		desc => "Remove edit if value on a nutriment field is equal to a value",
+		edit_rules => [{name => "Disallow ingredients", actions => [["ignore_if_equal_nutriment_sugar", 100]]},],
+		product => {"nutriment" => {"sugar_100g" => 22}},
+		form => {"nutriment_sugar" => 100},
+		result => 1,
+		delete_param => ["nutriment_sugar"],
+	},
+	{
+		id => "ignore_if_match_serving_size",
+		desc => "Remove edit if string value match a value",
+		edit_rules => [{name => "Disallow ingredients", actions => [["ignore_if_match_serving_size", "serving"]]},],
+		form => {"serving_size" => "serving"},
+		result => 1,
+		delete_param => ["serving_size"],
+	},
+	{
+		id => "ignore_if_regexp_match_brand",
+		desc => "Remove edit if string value match a regexp",
+		edit_rules =>
+			[{name => "Disallow ingredients", actions => [["ignore_if_regexp_match_brands", "(acme|hacky)"]]},],
+		form => {"brands" => "Another, Acme inc."},
+		result => 1,
+		delete_param => ["brands"],
+	},
+	{
+		id => "combine_actions",
+		desc => "Remove edit with combined actions",
+		edit_rules => [
+			{
+				name => "Disallow ingredients",
+				actions => [
+					["ignore_if_existing_ingredients_text_fr"],
+					["ignore_if_equal_nutriment_sugar", 100],
+					["ignore_if_match_serving_size", "serving"],
+					["ignore_if_regexp_match_brands", "(acme|hacky)"]
+				]
+			},
+		],
+		product => {ingredients_text_fr => "YES", "nutriment" => {"sugar_100g" => 22}},
+		form => {
+			ingredients_text_fr => "NOPE",
+			"nutriment_sugar" => 100,
+			"serving_size" => "serving",
+			"brands" => "Another, Acme inc."
+		},
+		result => 1,
+		delete_param => ["ingredients_text_fr", "ingredients_text", "nutriment_sugar", "serving_size", "brands"],
+	},
+	# FIXME: add tests on warning and slack notifications
+);
+
+my @edit_rules_backup = @edit_rules;
+
+# a global for fake CGI parameters
+my %form = ();
+
+sub fake_single_param ($name) {
+	return scalar $form{$name};
+}
+
+# removed params
+my @removed = ();
+
+sub fake_delete ($name) {
+	push @removed, $name;
+	return;
+}
+
+{
+	# monkey patch single_param
+	my $display_module = Test::MockModule->new('ProductOpener::Display');
+	$display_module->mock('single_param', \&fake_single_param);
+	# because this is a direct import in Products we have to monkey patch here, there also
+	my $products_module = Test::MockModule->new('ProductOpener::Products');
+	$products_module->mock('single_param', \&fake_single_param);
+	# patch GCI Delete
+	my $cgi_module = Test::MockModule->new('CGI');
+	$cgi_module->mock('Delete', \&fake_delete);
+	$products_module->mock('Delete', \&fake_delete);
+
+	foreach my $test_ref (@tests) {
+		# start a block to avoid side effect (edit_rules changes)
```

Comment:
I don't understand why this extra block changes something?

Reply:
Because I changed @edit_rules, and I want to be sure I restore them after, even if I got an error. To avoid my test eventually making other unrelated tests fails.

From my understanding whatever happens in the block, code after will be run normally (isn't it ?)

Reply:
Ok, I forgot the eval :-D